File: JavaScript/88_MergeSortedArray.js
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
var merge = function(nums1, m, nums2, n) {
    let end = m + n - 1;

    while (m > 0 && n > 0) {
        if (nums1[m - 1] >= nums2[n - 1]) {
            nums1[end] = nums1[m - 1];
            m--;
        } else {
            nums1[end] = nums2[n - 1]
            n--;
        }
        end--;
    }

    while (n > 0) {
        nums1[end] = nums2[n - 1];
        n--;
        end--;
    }
};

const testCases = [
    {
        nums1: [1,2,3,0,0,0],
        m: 3,
        nums2: [2,5,6],
        n: 3,
        expected: [1,2,2,3,5,6]
    },
    {
        nums1: [1],
        m: 1,
        nums2: [],
        n: 0,
        expected: [1]
    },
    {
        nums1: [0],
        m: 0,
        nums2: [1],
        n: 1,
        expected: [1]
    },
]

testCases.forEach((testCase, i) => {
    const output = testCase.nums1;
    merge(output, testCase.m, testCase.nums2, testCase.n);
    if (JSON.stringify(output) === JSON.stringify(testCase.expected)) {
        console.log(`Test Case ${i}: Passed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    } else {
        console.log(`Test Case ${i}: Failed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    }
});

File: CS/88_MergeSortedArray.cs
namespace LeetCodeSolutions {
    public static class MergeSortedArray {
        public static void Solution(int[] nums1, int m, int[] nums2, int n) {
            int end = m + n - 1;
            while (m > 0 && n > 0) {
                if (nums1[m - 1] >= nums2[n - 1]) {
                    nums1[end] = nums1[m - 1];
                    m--;
                } else {
                    nums1[end] = nums2[n - 1];
                    n--;
                }
                end--;
            }
            while (n > 0) {
                nums1[end] = nums2[n - 1];
                n--;
                end--;
            }
        }
        public static void RunTests() {
            Console.WriteLine("Running Length Of Merge Tests...");

            var testCases = new List<(int[], int, int[], int, int[])> {
                (
                    new int[] {1, 2, 3, 0, 0, 0}, 
                    3, 
                    new int[] {2, 5, 6},
                    3,
                    new int[] {1, 2, 2, 3, 5, 6}
                ),
                (
                    new int[] {1},
                    1,
                    new int[] {},
                    0,
                    new int[] {1}
                ),
                (
                    new int[] {0}, 
                    0, 
                    new int[] {1}, 
                    1,
                    new int[] {1}
                ),
            };

            foreach (var (nums1, m, nums2, n, expected) in testCases) {
                try {
                    int[] result = nums1;
                    
                    Solution(result, m, nums2, n);
                    Console.WriteLine(
                        $"Input: {string.Join(",", nums1)}, Merged List: {string.Join(",", nums2)} Vals: m {m},  n {n} " +
                        $"Expected: {string.Join(",", expected)} Result: {string.Join(",", result)}"
                    );
                } catch (Exception ex) {
                    Console.WriteLine($"Test Failed: {ex.Message}");
                }
            }
        }
    }
}